singularity-eos/eos/eos_davis.hpp

```diff
@@ -12,18 +12,159 @@
 // publicly and display publicly, and to permit others to do so.
 //------------------------------------------------------------------------------
 
+#ifndef _SINGULARITY_EOS_EOS_EOS_DAVIS_HPP_
+#define _SINGULARITY_EOS_EOS_EOS_DAVIS_HPP_
+
 #include <cmath>
+#include <cstdio>
+
+#include <singularity-eos/base/constants.hpp>
 #include <singularity-eos/base/root-finding-1d/root_finding.hpp>
-#include <singularity-eos/eos/eos.hpp>
+#include <singularity-eos/eos/eos_base.hpp>
 
 namespace singularity {
 
-PORTABLE_INLINE_FUNCTION Real square(const Real x) { return x * x; }
-PORTABLE_INLINE_FUNCTION Real cube(const Real x) { return x * x * x; }
-PORTABLE_INLINE_FUNCTION Real fourth(const Real x) { return x * x * x * x; }
-PORTABLE_INLINE_FUNCTION Real fifth(const Real x) { return x * x * x * x * x; }
+using namespace eos_base;
 
-constexpr Real onethird = 1.0 / 3.0;
+class DavisReactants : public EosBase<DavisReactants> {
+ public:
+  DavisReactants() = default;
+  PORTABLE_INLINE_FUNCTION
+  DavisReactants(const Real rho0, const Real e0, const Real P0, const Real T0,
+                 const Real A, const Real B, const Real C, const Real G0, const Real Z,
+                 const Real alpha, const Real Cv0)
+      : _rho0(rho0), _e0(e0), _P0(P0), _T0(T0), _A(A), _B(B), _C(C), _G0(G0), _Z(Z),
+        _alpha(alpha), _Cv0(Cv0) {}
+  DavisReactants GetOnDevice() { return *this; }
+  PORTABLE_INLINE_FUNCTION Real TemperatureFromDensityInternalEnergy(
```

Comment:
Maybe this is work for a different PR, but I think collocating declarations and implementations would be preferable.

Reply:
Yeah my preference is short declarations/implementations should be collocated, and longer ones separate. I did that for `IdealGas` and then quickly decided it was too tedious to do for all of them. I say we make that change piecewise as we work on individual EOS's.

Reply:
I created an issue: #170

Reply:
I ended up resolving this now.